hail/src/main/scala/is/hail/expr/types/physical/PCanonicalNDArray.scala

```diff
@@ -0,0 +1,211 @@
+package is.hail.expr.types.physical
+
+import is.hail.annotations.{ Region, StagedRegionValueBuilder, UnsafeOrdering}
+import is.hail.asm4s.{Code, MethodBuilder, _}
+
+final case class PCanonicalNDArray(elementType: PType, nDims: Int, required: Boolean = false) extends PNDArray  {
```

Comment:
The implementations are a straight copy paste, exception val modifiers on representation and derived properties